app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java

```diff
@@ -170,13 +300,85 @@ public void onClick(View v) {
 
         @Override
         public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
-            TokenCardMeta tcm = data.get(getAdapterPosition());
+            TokenCardMeta tcm = (TokenCardMeta) items.get(getAdapterPosition()).value;
             final Token t = tokensService.getToken(tcm.getChain(), tcm.getAddress());
             listener.onItemClick(t, isChecked);
         }
     }
 
+    public class LabelViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
+        TextView textTitle;
+        TextView textIgnore;
+
+        LabelViewHolder(View itemView) {
+            super(itemView);
+            textTitle = itemView.findViewById(R.id.text_title);
+            textIgnore = itemView.findViewById(R.id.text_ignore);
+            textIgnore.setOnClickListener(this);
+        }
+
+        @Override
+        public void onClick(View v) {
+            //TODO Implement Ignore click
+        }
+    }
+
     public interface ItemClickListener {
         void onItemClick(Token token, boolean enabled);
     }
+
+    @Override
+    public int getItemViewType(int position) {
+        return items.get(position).viewType;
+    }
+
+    @Override
+    public Filter getFilter() {
```

Comment:
Surely there's a simpler way to implement the search facility. So many flags and logic paths, it's impossible to work on. Can we use a different approach to achieve a more elegant method?

Reply:
I understand your concern but the complexity of the list is its Type. I need to manage the Single list with 4 parameters. 
1. Display Label, 2. Display Tokens, 3. Hidden Label and 4. Hidden Tokens

Searching leads to show hide views internally and switching between list made it too complex in the scenario like When a list in "Display" state, switched off, has to transfer to "Hidden" state and list as well. Such scenario was bit tedious to manage without flags and complex logic.

Reply:
Yes, understood that you need those differences in views; I'm just a bit concerned that there's a very long method with a lot of logic branches in it. This can be solved by having a separate viewholder for each of the different views. For example the title views can be a very small local class; instantly you've simplified the logic, probably sufficiently to eliminate a lot of the different paths through the method.

Let's do this and see what the code looks like after that

Reply:
I applied changes for above 2 points for Separate method and ViewHolders. About the filter method, I felt that it is necessary logic and it is at simplest. No doubt it seems longer, but it has only 2 logic points. 
1) When search string is blank, mark visible all views
2) When search string is present, compare string with name and mark invisible which are not in the list.

I used this approach so that we don't need to take any other "backup" array list which can be used to fill adapter list again. By simply using a flag to display item in the list works well without creating and dropping items inside the adapter. 

Could you please share you thoughts on this?